### ISSUE TITLE
Improve Errata management in W3C

### ISSUE THREAD

**Issue by dwsinger** [w3c/w3process#5]
Transferred from https://www.w3.org/community/w3process/track/issues/141
State: Open

**Comment by dwsinger** [comment]
Active at the AB as 'maintenance'

**Comment by jeffjaffe** [comment]
The AB continues to work on this issue.  Some of the proposals involve setting up different types of mechanisms to move forward with errata management which could interact with patent commitments, so the AB is getting advice from PSIG.

**Comment by plehegar** [comment]
Here is the full length proposal:

Note: orphan recommendations are Recommendations without an associated
   Working Group. (eg Web Crypto)

1. Facilitating maintenance:

      1. Improve Errata management
        a. Enforce a test driven approach
         "ALL substantive corrections are generally expected to have
          corresponding test(s) (such as a pull request in
          web-platforms-tests), either in the form of new tests or
          modifications to existing tests, or must include the rationale
          for why test updates are not required for the proposed update."

          -> update the Process:
          [[
             Readers of the Recommendation should be able easily to find
             and see the errata that apply to that specific Recommendation
             with their associated tests.
          ]]
          https://www.w3.org/2017/06/maintenance_proposal.html#errata

        b. Automate the errata pages by using issues, commit, and testing
          information

          -> Modern tooling

       (ETA: First version/prototype: November 2017)

        c. Use GitHub labels for differentiate between editorial and
           substantive changes (as well as horizontal review labels)

      2. Track issues for ALL active Recommendations

        a. Each Recommendation must have an associated github public
           repository to track editor drafts and issues
          -> update publication and transition requirements
          -> create github repositories for existing recommendations
             and them to indicate so
          -> all commits must be done through reviewed pull requests
          -> No IP manager for repositories with orphan Recommendations

        b. Develop a *public* tracking and reporting system for issues,
           errata pages, and editor' drafts
          "Such error reports should be processed no less frequently than
          quarterly."
          https://www.w3.org/2017/Process-20170301/#errata
          -> Modern tooling

       (ETA: First version/prototype: September 2017)

        c. Appoint a maintenance contact in charge of receiving and
           tracking requests for orphan Recommendations
           (Note: this may not be someone from W3C Staff)

          -> update publication and transition requirements
          -> this individual becomes the default reviewer for all of the
             github repositories for orphan Recommendations (it may be
             possible to add to that default if someone else is more
             appropriate)

      3. Provide maintenance BP Guidelines (including obsolescence and
         rescinding)

        Examples of editorial and substantive corrections
        Who to contact?
        Using GitHub to update a document
        Clarify wide review requirements for errata pages
         https://www.w3.org/2017/Process-20170301/#wide-review
         * Are errata published on errata pages for at least one month
           enough to qualify as wide review?
         * Should we apply CR exit criteria early on, such as having 2 or
           more implementations?
         * Should we announce new substantive corrections to
           public-review-announce@w3.org?

         -> new documentation (editor is maintenance contact)

      4. when a Working Group closes, consider one of the following
         options:
         a. Reassign the now-orphaned Recommendations to an existing or
            new group (preferred) (eg SVG)
         b. Extend end-of-life working group with a Maintenance mode (less
            suitable option of all)
         c. Request that W3C publish an updated document when needed (eg
            Web Crypto, Social Web-*, etc.)

        To enable c), we need to allow orphan Recommendations to be
        maintained by non-W3C Working Groups (such as CG) or any
        individual.

         -> process update:
          "To make substantive corrections to a Recommendation, or where 
there were votes against publishing the corrections directly as a 
Recommendation, a Working Group may request publication of a Candidate 
Recommendation or W3C may publish a Candidate Recommendation, without 
passing through earlier maturity levels."

          https://www.w3.org/2017/06/maintenance_proposal.html#revised-rec

          Reminder for c):
            * Despite the lack of IP commitment, we're *only* talking
              about substantive corrections here. No new features!
              It's better to have a maintained spec with no IP commitments
              for its substantive corrections than an old specification.
            * Because it's substantive corrections, the Process will
              trigger a Candidate Recommendation. So the Advisory
              Committee still gets to speak up at Proposed Recommendation
              stage (or earlier).

           For a) and b), we would not require the individual to be part
           of the Working Group if one would like to submit proposed
           substantive corrections.

          -> consider deactivating the IP manager for those

          Example of Amended Recommendation:
          [[
Micropub
W3C Recommendation 21 Juin 2018 (Amended by W3C)
[...]

The Recommendation from which this document is derived was produced by a 
group operating under the 5 February 2004 W3C Patent Policy. W3C 
maintains a public list of any patent disclosures made in connection 
with the deliverables of the group; that page also includes instructions 
for disclosing a patent. An individual who has actual knowledge of a 
patent which the individual believes contains Essential Claim(s) must 
disclose the information in accordance with section 6 of the W3C Patent 
Policy. This Amended Recommendation was produced by incorporating errata 
after that group closed.

          ]]
          https://www.w3.org/2017/06/Micropub.html

2.  Empower Individuals:

   1. The maintenance status of existing Recommendations must be publicly
available. For each recommendation, it must states:
     - the Working Group in charge of maintaining the Recommendation, if
       any
     - The URL of the Recommendation and its editor's draft
     - The URL of the errata page
     - The URL of the GitHub issues list
     - The date of the most recent commit to the editor's draft/errata
       pages
     - The number of pending open issues
     - The number of pending open pull requests
     - If there are open issues and no updates were done in them within
       last 6 months, it should suggest Individuals to reach out to the
       maintenance contact to become an editor

    You should be able to sort through those statuses.

    -> Modern tooling
       (ETA: First version/prototype: September 2017)

   2. Edited Recommendation must list the individuals who contributed to
      the maintenance of the Recommendation

   3. The maintenance contact should write a yearly public report on
      maintenance status, asking for and thanking volunteers

   4. Facilitate transition requests for Edited Recommendation
      * Automatic generation of transition requests from errata
        pages?
             (Note that we're already looking at switching to GitHub
              issues to facilitate transition requests in general...)

    -> Modern tooling

       (ETA: First version/prototype: December 2017)

**Comment by chaals** [comment]
Fixed for now by #61. I'll remove the 2018 label but leave this open and explain why I think we still have work to do later...

**Comment by dwsinger** [comment]
please don't remove labels which are owned by the CG...this represents a consensus on what's a candidate for the process next year. maybe we still have work to do, maybe we won't succeed in landing it. but we do need to track it as a candidate

**Comment by dontcallmedom** [comment]
The text in the proposed new process document says that the amended Recommendation process is initiated by "the W3C", but I don't understand what that means in practice. Is it the staff? The Director? I think this would benefit from being clarified

**Comment by dwsinger** [comment]
...I think the implication is that somehow as a community we realize the need. I agree, I think it has to the team that is going to initiate the process steps -- can you, do you wish to, suggest a specific edit to clarify this?

**Comment by jeffjaffe** [comment]
I'm not sure where in the document you are asking your question.  I would expect that there would be circumstances where a CG is handling maintenance and requests an Amended Recommendation.

**Comment by dontcallmedom** [comment]
the specific text I'm referring to is
> To publish a Candidate recommendation, [...], in the case of a document intended to become an Amended Recommendation, **the W3C** MUST show…
https://w3c.github.io/w3process/#candidate-rec

(similar text also exists in the "Proposed Recommendation" section)

In practice, it is unclear with that text how one can suggest triggering the Amended Recommendation process, and when it is done, who is responsible to see it through.

@plehegar 's original proposal seems to suggest that the "how" relies on a designated maintenance contact, but I think the process should be clearer on the rights and obligations around this amended recommendation process.

It could be something like:
* the W3C Management MUST designate a Maintenance Contact for Recommendations no longer maintained by an active Working Group
* The Maintenance Contact MUST start the Amended Recommendation process no later than X weeks/months after a correction that produces substantive changes but does not add new features has been submitted, provided that evidence that the said correction would meet the Candidate Recommendation criteria can be gathered

**Comment by jeffjaffe** [comment]
I'm not overly in favor of Dom's second bullet as it appears to overmanage the situation.  I don't think we know that every substantive correction is so impactful that it requires MUST start the Amended process.

The first bullet is interesting.  I am curious what is implied by the Maintenance Contact.  Is that also a person?; always a team member?; could it be a CG?

**Comment by dwsinger** [comment]
I agree with Jeff, I'd rather be silent on the second bullet. On the first, I am fine with saying "whoever you are -- member, CG member, public person, the devil incarnate, whatever -- find a team member to be the shepherd/contact of this Amended Rec." I don't think "must" works, as I can't tell when the must kicks in. State as a fact? "The request for an amended rec is made by a member of the team" (possibly inspired by someone else)

**Comment by nigelmegitt** [comment]
Two comments from me on this:

1. Having two things that are synonymous but intended to have different connotations seems like a recipe for confusion: 
   *   if the distinction between an Edited Recommendation and an Amended Recommendation is that the former was made by a WG and the latter was not (as described in §6.7.2), then firstly, I question the utility of making this distinction, and secondly, can we find a better name reflecting _who_ made the change? Maybe "Edited Recommendation (non-Chartered)"?
   *  Alternatively if the distinction is the class of substantive change that is acceptable (as described in §6.1.2) then can we find a better name reflecting that? Maybe "Clarified Recommendation", or "Edited Recommendation (Practice-aligned)".
2. I don't understand how Wide Review can practically be achieved in the absence of a WG. How can any resolutions be made or dispositions sensibly be approved? The actual set of people involved would be a staff contact and the change requester. I'd worry a little that in principle this process could be gamed or hijacked, as follows:
   * Member objects to rechartering WG.
   * WG is closed.
   * Member proposes a change, on the basis that it is not a new feature, but that it is a substantive improvement.
   * Wide review is sought, but between the staff contact and the change proposer, no consensus is achieved to respond to any of the review comments with a further change. "wontfix" is fed back to all comments. Some commenters are grumpy, but they have only further one chance to comment, and that comment is duly recorded.
   * Request to transition is made to Director, who sees that wide review has been sought and all comments have been disposed, but that some were grumpy, but has no real reason not to approve.

I'd like to know if the process team considered a mini WG charter process instead for dealing with this kind of change? The idea would be that a member proposes to charter a WG for the sole purpose of making a maintenance update of a Recommendation, being an Edited Recommendation (or even a "Clarified Recommendation"), with the proviso that the the scope of change excludes any new features. I'd suggest that such a charter shall impose a short timescale lifetime for the temporary WG, e.g. 3 months, and that the default would be to approve it, assuming that a staff contact can be resourced. Obviously this would not be available unless the original WG had closed. The advantage of this approach would be that other Members would have the opportunity to join and participate in the usual way. Staff contact resourcing levels could be managed by setting a maximum number of temporary WGs that can simultaneously be active and queuing the remainder in FIFO order.

**Comment by dwsinger** [comment]
Yes, we are concerned about the plethora of similar names, but we didn't want to fiddle with names that previously existed (at least, not without dialogue). Personally, I would call "edited recommendations" just "recommendations" since they are a WG product.

Wide review is done by everyone *outside* the WG, so we want to at least consider the possibility that a bug fix might introduce security, privacy, internationalization, accessibility, or other issues, and catch them. If it does, we may have to think again. Similarly, the point of having an AC vote is to allow the AC to say "hang on, the WG explicitly rejected that when it existed" and so on.

Yes, we briefly considered forming a WG for a bug fix, but the conclusion was that few if any would join in order to get under a patent policy umbrella over the entire spec. for a bug fix; I suspect many legal departments would say it wasn't worth their time.

Basically, if anyone on the AC says "guys, this is not a simple question, we need a WG" we will probably have to think again.

**Comment by nigelmegitt** [comment]
> Wide review is done by everyone _outside_ the WG

Yes, my point was about how to handle any wide review feedback received if there are only two people involved, and no mechanism for allowing anyone else to participate.

**Comment by dwsinger** [comment]
I think if the wide review throws up problems, the team member and AC are going to have to make a call on how/whether to proceed. "Don't do it", "adjust the fix", "form a WG", at least come to mind.

**Comment by plehegar** [comment]
regarding the terms, note that "Edited" and "Amended" are mostly only terms for the Process. See the sample
https://www.w3.org/2017/06/Micropub.html for example of what an Amended Recommendation would like.

**Comment by dwsinger** [comment]
The lingering concern is that the process might be seen as implying that we need full errata management for superceded specs. The thought is that though you should be able to report an erratum on an old spec (e.g. version 3, when version 5 is the current, and 3 is marked superceded) the WG may well do a quick triage and see whether it also applies to the latest version and drop it if not. So, the ability to track errata is a mandate for old specs; fixing them should be something the WG can choose to do, but it should not feel under any obligation other than to see if the problem exists in the current version. This is not Process2018 material; we leave the issue open, however, for future consideration.

**Comment by dwsinger** [comment]
This seems a nit; clarify that errata management for superceded specs can be restricted to transferring still-alive bugs to the current version.
Is this a priority?

**Comment by jeffjaffe** [comment]
Let's see how amended and superseded work out before we do more in this area.

**Comment by dwsinger** [comment]
AB thinks we should watch the current obsolete and amended rec processes, and improve practice, and close this pending a specific process issue request
